### PR TITLE
[PDI-3314] - very slow database lookup if "not all equal conditions" and "load all rows at start"

### PR DESCRIPTION
**Pull request by akhayrutdinov**
@brosander, review it please

### CODE REVIEW

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Dpentaho.coding.standards.version=1.0.1 -Djunit.jvmargs=-XX:PermSize=256m -f engine/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupData.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DefaultCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/Index.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContext.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupUTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IndexTestBase.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/NeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCacheTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContextTest.java


Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards

File engine/src/org/pentaho/di/trans/steps/databaselookup/DefaultCache.java
1. Message: 'switch' is not followed by whitespace. at Line: 87 Column: 19









Unit test coverage change:
============
org.pentaho.di.trans.steps.randomvalue.RandomValueMeta:
 * instructionPercentCoverageChange : +0.15%
 * branchPercentCoverageChange : +2.08%
 * complexityPercentCoverageChange : +2.13%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.IsNullIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :94.12%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :95.52%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :89.47%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :92.59%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache:
 * branchPercentCoverage :97.14%
 * linePercentCoverage :97.44%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :96.00%
 * instructionPercentCoverage :98.35%
    
org.pentaho.di.trans.steps.excelwriter.ExcelWriterStepMeta:
 * instructionPercentCoverageChange :**-0.08%**
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexValueComparator:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :75.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :80.65%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.GtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookup:
 * instructionPercentCoverageChange : +6.58%
 * branchPercentCoverageChange : +9.86%
 * linePercentCoverageChange : +9.79%
 * complexityPercentCoverageChange : +10.79%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.SearchingContext:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache.Builder:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.uniquerows.UniqueRowsMeta:
 * instructionPercentCoverageChange : +0.17%
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookupMeta:
 * instructionPercentCoverageChange : +2.22%
 * linePercentCoverageChange : +4.01%
 * complexityPercentCoverageChange : +3.92%
 * methodPercentCoverageChange : +7.02%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.new Comparator() {...}:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.LtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.sortedmerge.SortedMergeMeta:
 * instructionPercentCoverageChange : +0.18%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.EqIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexedValue:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.DefaultCache:
 * branchPercentCoverage :25.00%
 * linePercentCoverage :42.11%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :17.50%
 * instructionPercentCoverage :40.22%

**Comment by akhayrutdinov** [comment]
I've fixed the violation

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Dpentaho.coding.standards.version=1.0.1 -Djunit.jvmargs=-XX:PermSize=256m -f engine/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupData.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DefaultCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/Index.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContext.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupUTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IndexTestBase.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/NeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCacheTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContextTest.java










Unit test coverage change:
============
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache.Builder:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache:
 * branchPercentCoverage :97.14%
 * linePercentCoverage :97.44%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :96.00%
 * instructionPercentCoverage :98.35%
    
org.pentaho.di.trans.step.StepMeta:
 * instructionPercentCoverageChange :**-0.19%**
 * linePercentCoverageChange :**-0.27%**
    
org.pentaho.di.job.entries.sql.JobEntrySQL:
 * instructionPercentCoverageChange :**-1.64%**
 * linePercentCoverageChange :**-1.97%**
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexedValue:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.excelwriter.ExcelWriterStepMeta:
 * instructionPercentCoverageChange :**-0.08%**
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexValueComparator:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :75.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :80.65%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.GtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookupMeta:
 * instructionPercentCoverageChange : +2.22%
 * linePercentCoverageChange : +4.01%
 * complexityPercentCoverageChange : +3.92%
 * methodPercentCoverageChange : +7.02%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.new Comparator() {...}:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.SearchingContext:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.replacestring.ReplaceStringMeta:
 * instructionPercentCoverageChange : +0.07%
    
org.pentaho.di.trans.steps.uniquerows.UniqueRowsMeta:
 * instructionPercentCoverageChange :**-0.17%**
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookup:
 * instructionPercentCoverageChange : +6.58%
 * branchPercentCoverageChange : +9.86%
 * linePercentCoverageChange : +9.79%
 * complexityPercentCoverageChange : +10.79%
    
org.pentaho.di.repository.kdr.KettleDatabaseRepositoryCreationHelper:
 * instructionPercentCoverageChange :**-2.43%**
 * branchPercentCoverageChange :**-1.64%**
 * linePercentCoverageChange :**-2.95%**
 * complexityPercentCoverageChange :**-0.69%**
    
org.pentaho.di.trans.steps.databaselookup.DefaultCache:
 * branchPercentCoverage :25.00%
 * linePercentCoverage :42.11%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :17.50%
 * instructionPercentCoverage :40.22%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.IsNullIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :94.12%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :95.52%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :89.47%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :92.59%
    
org.pentaho.di.trans.steps.sortedmerge.SortedMergeMeta:
 * instructionPercentCoverageChange : +0.18%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.LtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.textfileinput.TextFileInputMeta:
 * branchPercentCoverageChange :**-0.53%**
    
org.pentaho.di.trans.steps.databaselookup.readallcache.EqIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%

**Comment by akhayrutdinov** [comment]
@brosander, review it please

File: engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java [hunk 1]
```diff
@@ -77,7 +71,8 @@ public DatabaseLookup( StepMeta stepMeta, StepDataInterface stepDataInterface, i
    * @return the resulting row after the lookup values where added
    * @throws KettleException In case something goes wrong.
    */
-  private synchronized Object[] lookupValues( RowMetaInterface inputRowMeta, Object[] row ) throws KettleException {
+  // package-local visibility for testing purposes
```
Comment (brosander) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @akhayrutdinov can we use @VisibleForTesting instead?

Comment (akhayrutdinov) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @brosander, ha-ha-ha :smile: 
Once I was told not to use it henceforth (I remember @dkincade was involved in that conversation). So, what is correct?

File: engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java [hunk 2]
```diff
@@ -640,6 +479,60 @@ private void loadAllTableDataIntoTheCache() throws KettleException {
...
+  }
+
+  private void putToReadOnlyCache( Database db, List<Object[]> rows ) {
+    ReadAllCache.Builder cacheBuilder = new ReadAllCache.Builder( data, rows.size() );
+    RowMetaInterface returnRowMeta = db.getReturnRowMeta();
+
+    // all keys have the same row meta,
+    // it is useless to re-create it each time
+    final int keysAmount = meta.getStreamKeyField1().length;
+    RowMeta keyMeta = new RowMeta();
```
Comment (brosander) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @akhayrutdinov please consider using RowMeta.clone() as it should now be much faster than adding individual value metas (https://github.com/pentaho/pentaho-kettle/commit/88e0ebbfb0caa67c5f72d19066697d1fa1856793)

File: engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java [hunk 3]
```diff
@@ -640,6 +479,60 @@ private void loadAllTableDataIntoTheCache() throws KettleException {
     }
   }
 
+  private void putToDefaultCache( Database db, List<Object[]> rows ) {
+    RowMetaInterface returnRowMeta = db.getReturnRowMeta();
+    // Copy the data into 2 parts: key and value...
+    //
+    for ( Object[] row : rows ) {
+      int index = 0;
+      RowMeta keyMeta = new RowMeta();
```
Comment (brosander) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @akhayrutdinov please consider using RowMeta.clone() as it should now be much faster than adding individual value metas (https://github.com/pentaho/pentaho-kettle/commit/88e0ebbfb0caa67c5f72d19066697d1fa1856793)

File: engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java (hunk 1 above)
Comment (brosander) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @akhayrutdinov I will find out

Comment (brosander) [inline engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java]: @akhayrutdinov I spoke with Kincade and asked around and the general consensus is if guava is already a dependency, @VisibleForTesting is preferred

**Comment by dkincade** [comment]
If I remember correctly, that was more of a conversation about using default scope vs protected. I don't remember asking people to use/not use @VisibleForTesting ... it looks like it is used a number of places and the dependency is already in Kettle. So we should be good to use it.

**Comment by akhayrutdinov** [comment]
@brosander, I've updated the PR. Review it please.

1. Added `@VisibleForTesting`
2. Replaced manual RowMeta populating with `clone()`

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Dpentaho.coding.standards.version=1.0.1 -Djunit.jvmargs=-XX:PermSize=256m -f engine/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookup.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupData.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/DefaultCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/Index.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndex.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCache.java
  engine/src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContext.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/DatabaseLookupUTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/EqIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/GtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IndexTestBase.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/IsNullIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/LtIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/NeIndexTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/ReadAllCacheTest.java
  engine/test-src/org/pentaho/di/trans/steps/databaselookup/readallcache/SearchingContextTest.java










Unit test coverage change:
============
org.pentaho.di.trans.steps.textfileinput.TextFileInputMeta:
 * branchPercentCoverageChange : +0.53%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.IsNullIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :94.12%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :95.52%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :89.47%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :92.59%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache:
 * branchPercentCoverage :97.14%
 * linePercentCoverage :97.44%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :96.00%
 * instructionPercentCoverage :98.35%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexValueComparator:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :75.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :80.65%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.GtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.sortedmerge.SortedMergeMeta:
 * instructionPercentCoverageChange :**-0.18%**
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookup:
 * instructionPercentCoverageChange : +6.11%
 * branchPercentCoverageChange : +9.34%
 * linePercentCoverageChange : +9.36%
 * complexityPercentCoverageChange : +10.20%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.SearchingContext:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.repository.kdr.KettleDatabaseRepositoryCreationHelper:
 * instructionPercentCoverageChange :**-2.43%**
 * branchPercentCoverageChange :**-1.64%**
 * linePercentCoverageChange :**-2.95%**
 * complexityPercentCoverageChange :**-0.69%**
    
org.pentaho.di.trans.step.StepMeta:
 * instructionPercentCoverageChange :**-0.19%**
 * linePercentCoverageChange :**-0.27%**
    
org.pentaho.di.trans.steps.databaselookup.DatabaseLookupMeta:
 * instructionPercentCoverageChange : +2.22%
 * linePercentCoverageChange : +4.01%
 * complexityPercentCoverageChange : +3.92%
 * methodPercentCoverageChange : +7.02%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.new Comparator() {...}:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.ReadAllCache.Builder:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.LtIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.core.auth.core.impl.DefaultAuthenticationConsumerFactory:
 * branchPercentCoverageChange :**-7.14%**
 * complexityPercentCoverageChange :**-7.69%**
    
org.pentaho.di.trans.steps.databaselookup.readallcache.EqIndex:
 * branchPercentCoverage :100.00%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.readallcache.Index.IndexedValue:
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :100.00%
 * instructionPercentCoverage :100.00%
    
org.pentaho.di.trans.steps.databaselookup.DefaultCache:
 * branchPercentCoverage :25.00%
 * linePercentCoverage :42.11%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :17.50%
 * instructionPercentCoverage :40.22%

**Pull request merged**